### PR TITLE
[MoE Refactor][17/N] Apply Refactor to Bf16

### PR DESCRIPTION
**Pull request by zyongye**
## Test Plan

gsm8k result for Triton kernel, flashinfer_cutlass kernel and aiter rocm kernel for `Qwen/Qwen3-30B-A3B` in TP(triton), TEP (flashinfer cutlass) and rocm. 

```
lm_eval \
                --model local-completions \
                --tasks gsm8k \
                --model_args "model=Qwen/Qwen3-30B-A3B,base_url=http://localhost:8000/v1/completions,num_concurrent=1000,tokenized_requests=False"
```

## Test Result
Triton

```
|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value |   |Stderr|
|-----|------:|----------------|-----:|-----------|---|-----:|---|-----:|
|gsm8k|      3|flexible-extract|     5|exact_match|↑  |0.8643|±  |0.0094|
|     |       |strict-match    |     5|exact_match|↑  |0.9045|±  |0.0081|
```

Flashinfer

```
|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value |   |Stderr|
|-----|------:|----------------|-----:|-----------|---|-----:|---|-----:|
|gsm8k|      3|flexible-extract|     5|exact_match|↑  |0.8544|±  |0.0097|
|     |       |strict-match    |     5|exact_match|↑  |0.8984|±  |0.0083|
```

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [ ] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [ ] The test plan, such as providing test command.
- [ ] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
- [ ] (Optional) Release notes update. If your change is user facing, please update the release notes draft in the [Google Doc](https://docs.google.com/document/d/1YyVqrgX4gHTtrstbq8oWUImOyPCKSGnJ7xtTpmXzlRs/edit?tab=t.0).
</details>

---

> [!NOTE]
> <sup>[Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) is generating a summary for commit f354ef2eabe320528a5815978bc8b87ee9fb505d. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

---

> [!NOTE]
> Refactors unquantized MoE to centralize backend selection and kernel setup; adds BF16 GSM8K eval configs for MoE models and updates test config lists.
> 
> - Introduces `UnquantizedMoeBackend` enum with `select_unquantized_moe_backend` and `convert_to_unquantized_kernel_format`, consolidating Aiter/FlashInfer/Triton selection, weight shuffling (`swap_w13_to_w31`), and kernel init via `_setup_kernel` in `unquantized_fused_moe_method.py`.
> - Simplifies `process_weights_after_loading` and `maybe_make_prepare_finalize`; adjusts DP/EP checks (uses `dp_rank > 1`) and logs; preserves inplace behavior differences per backend.
> - Adds BF16 GSM8K configs for `Llama-4-Scout`, `Mixtral-8x7B`, and `Qwen3-30B-A3B` (both `fi-cutlass` and `triton`) and registers them in `config-b200.txt` and `config-h100.txt`.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit f354ef2eabe320528a5815978bc8b87ee9fb505d. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

---

> [!NOTE]
> Streamlines unquantized MoE execution and adds BF16 evaluation configs.
> 
> - Refactors unquantized MoE: introduces `vllm/.../oracle/unquantized.py` with `UnquantizedMoeBackend`, `select_unquantized_moe_backend`, `convert_to_unquantized_kernel_format`, and `make_unquantized_moe_kernel`; updates `unquantized_fused_moe_method.py` to use centralized backend selection, DP/EP checks (`dp_rank > 1`), weight shuffling/swapping, and kernel setup; simplifies `maybe_make_prepare_finalize` and CPU/XPU handling; adds kernel assertion.
> - Adds BF16 GSM8K configs for `Llama-4-Scout`, `Mixtral-8x7B`, and `Qwen3-30B-A3B` (Triton and FlashInfer CUTLASS variants with `VLLM_USE_FLASHINFER_MOE_FP16` for CUTLASS) and registers them in `config-b200.txt` and `config-h100.txt`.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit f5fe3788a4f4933c64046e5d6f338ddce5c0fc16. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

---

> [!NOTE]
> Streamlines unquantized MoE and expands BF16 test coverage.
> 
> - New `oracle/unquantized.py`: defines `UnquantizedMoeBackend`, backend selection (`select_unquantized_moe_backend`), weight layout conversion, and kernel construction
> - Refactors `unquantized_fused_moe_method.py` to use centralized backend logic, simplify `process_weights_after_loading`/`maybe_make_prepare_finalize`, adjust DP/EP checks, and assert kernel presence
> - Adds BF16 GSM8K configs (Triton and FlashInfer CUTLASS variants) for `Llama-4-Scout`, `Mixtral-8x7B`, `Qwen3-30B-A3B`; registers them in `config-b200.txt` and `config-h100.txt`
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit bfe742bc808327197c0c05c7a6c98cfa65c35f73. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

---

> [!NOTE]
> <sup>[Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) is generating a summary for commit 3759b5ff799a6b35dbba16a42b8ef85671fb1e8e. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

---

> [!NOTE]
> Streamlines unquantized MoE execution and expands BF16 test coverage.
> 
> - New `oracle/unquantized.py` defines `UnquantizedMoeBackend`, backend selection (`select_unquantized_moe_backend`), weight layout conversion, and kernel construction for AITER, FlashInfer CUTLASS, and Triton
> - Refactors `unquantized_fused_moe_method.py` to use centralized backend logic, simplifying `maybe_make_prepare_finalize`, CPU/XPU handling, and kernel setup; adds weight shuffling/swapping and kernel assertion
> - Adds BF16 GSM8K configs (Triton and FlashInfer CUTLASS variants) for `Llama-4-Scout`, `Mixtral-8x7B`, and `Qwen3-30B-A3B`, plus a DP+EP Triton config for `Qwen3-30B-A3B`; registers them in `config-b200.txt` and `config-h100.txt`
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 522cb1c327bae5b46778a18f03f616f68ec582b3. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

---

> [!NOTE]
> <sup>[Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) is generating a summary for commit e0752934dec45fd6c34520fcc16e71254c84615d. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

### CODE REVIEW

File: vllm/model_executor/layers/fused_moe/unquantized_fused_moe_method.py [hunk 1]
```diff
@@ -55,43 +58,70 @@
...
-        )
-        if self.flashinfer_cutlass_moe_enabled:
+class UnquantizedMoeBackend(Enum):
+    NONE = 0
+    FLASHINFER_CUTLASS = 1
+    AITER = 2
+    TRITON = 3
+
+
+def get_unquantized_moe_backend(
```
Comment (robertgshaw2-redhat) [inline vllm/model_executor/layers/fused_moe/unquantized_fused_moe_method.py]: `get` --> `select` is the new convention

File: vllm/model_executor/layers/fused_moe/unquantized_fused_moe_method.py [hunk 2]
```diff
@@ -55,43 +58,70 @@
...
+    return UnquantizedMoeBackend.NONE
+
+
+@CustomOp.register("unquantized_fused_moe")
+class UnquantizedFusedMoEMethod(FusedMoEMethodBase, CustomOp):
+    """MoE method without quantization."""
+
+    def __init__(self, moe: FusedMoEConfig):
+        super().__init__(moe)
+        self.unquantized_backend = get_unquantized_moe_backend(
```
Comment (robertgshaw2-redhat) [inline vllm/model_executor/layers/fused_moe/unquantized_fused_moe_method.py]: rather than passing the moe parallel config, let's just pass use_dp and use_ep

File: vllm/model_executor/layers/fused_moe/unquantized_fused_moe_method.py [hunk 3]
```diff
@@ -197,6 +227,63 @@ def _maybe_pad_weight(self, weight: torch.Tensor) -> torch.Tensor:
...
+                    ep_rank=self.moe.moe_parallel_config.ep_rank,
+                    ep_size=self.moe.moe_parallel_config.ep_size,
+                ),
+            )
+            self.use_inplace = False
+        elif self.unquantized_backend == UnquantizedMoeBackend.AITER:
+            self.kernel = mk.FusedMoEModularKernel(
+                MoEPrepareAndFinalizeNoEP(),
+                AiterExperts(self.moe_quant_config),
+                shared_experts=None,
```
Comment (robertgshaw2-redhat) [inline vllm/model_executor/layers/fused_moe/unquantized_fused_moe_method.py]: drop the shared_experts, this is the default

File: vllm/model_executor/layers/fused_moe/unquantized_fused_moe_method.py [hunk 4]
```diff
@@ -246,45 +333,10 @@ def process_weights_after_loading(self, layer: torch.nn.Module) -> None:
...
-                    ),
-                )
-            else:
-                self.use_inplace = True
-                self.kernel = mk.FusedMoEModularKernel(
-                    MoEPrepareAndFinalizeNoEP(),
-                    TritonExperts(self.moe_quant_config),
-                    shared_experts=None,
-                )
+            self._convert_weights_to_kernel_format(layer=layer)
```
Comment (robertgshaw2-redhat) [inline vllm/model_executor/layers/fused_moe/unquantized_fused_moe_method.py]: to keep things consistent, there should be a single function called `_setup_kernel()` which does these two steps

File: vllm/model_executor/layers/fused_moe/unquantized_fused_moe_method.py [hunk 5]
```diff
@@ -197,6 +227,63 @@ def _maybe_pad_weight(self, weight: torch.Tensor) -> torch.Tensor:
...
+                MoEPrepareAndFinalizeNoEP(),
+                TritonExperts(self.moe_quant_config),
+                shared_experts=None,
+            )
+        elif self.unquantized_backend == UnquantizedMoeBackend.NONE:
+            raise ValueError(
+                "Unable to select quantization backend, please check supported backend."
+            )
+
+    def _convert_weights_to_kernel_format(
```
Comment (robertgshaw2-redhat) [inline vllm/model_executor/layers/fused_moe/unquantized_fused_moe_method.py]: this function should be in `oracle` (we will eventually make it a method of the Expert)

Just the "replace_parameter" should be part of this function

File: tests/evals/gsm8k/configs/moe-refactor/Llama-4-Scout-BF16-triton.yaml [hunk 6]
```diff
@@ -0,0 +1,6 @@
+model_name: "meta-llama/Llama-4-Scout-17B-16E-Instruct"
+accuracy_threshold: 0.92
+num_questions: 1319
+num_fewshot: 5
+server_args: "--enforce-eager --max-model-len 8192 --tensor-parallel-size 8"
```
Comment (robertgshaw2-redhat) [inline tests/evals/gsm8k/configs/moe-refactor/Llama-4-Scout-BF16-triton.yaml]: you need to update this to tp=2

File: vllm/model_executor/layers/fused_moe/unquantized_fused_moe_method.py [hunk 7]
```diff
@@ -52,43 +54,90 @@
...
+
+@CustomOp.register("unquantized_fused_moe")
+class UnquantizedFusedMoEMethod(FusedMoEMethodBase, CustomOp):
+    """MoE method without quantization."""
+
+    def __init__(self, moe: FusedMoEConfig):
+        super().__init__(moe)
+        self.unquantized_backend = select_unquantized_moe_backend(
+            use_ep=self.moe.moe_parallel_config.use_ep,
+            use_dp=self.moe.moe_parallel_config.dp_rank > 1,
```
[the remark this answers is not in the document]
Comment (robertgshaw2-redhat) [reply vllm/model_executor/layers/fused_moe/unquantized_fused_moe_method.py]: the bot is correct.

Can you add a dp/ep test?

File: tests/evals/gsm8k/configs/moe-refactor/Llama-4-Scout-BF16-triton.yaml (hunk 6 above)
Comment (zyongye) [reply tests/evals/gsm8k/configs/moe-refactor/Llama-4-Scout-BF16-triton.yaml]: The llama4 is 108B parameter and can't fit in 2 H200 gpus.

File: vllm/model_executor/layers/fused_moe/oracle/unquantized.py [hunk 8]
```diff
@@ -0,0 +1,161 @@
...
+            backend = UnquantizedMoeBackend.FLASHINFER_CUTLASS
+        else:
+            if use_ep and (not use_dp):
+                logger.info_once(
+                    "FlashInfer CUTLASS MoE is available for EP"
+                    " but not enabled, consider setting"
+                    " VLLM_USE_FLASHINFER_MOE_FP16=1 to enable it.",
+                    scope="local",
+                )
+            elif use_dp:
```
Comment (robertgshaw2-redhat) [inline vllm/model_executor/layers/fused_moe/oracle/unquantized.py:84]: Why does this kernel work with TP/EP but not DP/EP?

I dont see why there would be a distinction. I actually think this should work fine with the MK structure

Comment (zyongye) [reply vllm/model_executor/layers/fused_moe/oracle/unquantized.py:84]: It needs further investigation. The original code says it doesn't support DP/EP.

Comment (zyongye) [reply vllm/model_executor/layers/fused_moe/oracle/unquantized.py:84]: Oh maybe it's just not in `selec_gemm_impl` yet.

File: tests/evals/gsm8k/configs/moe-refactor/config-b200.txt [hunk 9]
```diff
@@ -11,3 +11,9 @@ Qwen3-30B-A3B-NvFp4-ModelOpt-marlin.yaml
 Qwen3-30B-A3B-NvFp4-ModelOpt-fi-trtllm.yaml
 Qwen3-30B-A3B-NvFp4-ModelOpt-fi-cutlass.yaml
 Qwen3-30B-A3B-NvFp4-ModelOpt-fi-cutlass-dp-ep.yaml
+Llama-4-Scout-BF16-fi-cutlass.yaml
```
Comment (robertgshaw2-redhat) [inline tests/evals/gsm8k/configs/moe-refactor/config-b200.txt:14]: run half on the b200 and some on the h100 for CI time / budget

File: tests/evals/gsm8k/configs/moe-refactor/Llama-4-Scout-BF16-triton.yaml (hunk 6 above)
Comment (zyongye) [reply tests/evals/gsm8k/configs/moe-refactor/Llama-4-Scout-BF16-triton.yaml]: Move llama4 tests to b200.

File: tests/evals/gsm8k/configs/moe-refactor-dp-ep/Qwen3-30B-A3B-BF16-triton.yaml [hunk 10]
```diff
@@ -0,0 +1,5 @@
+model_name: "Qwen/Qwen3-Coder-30B-A3B"
+accuracy_threshold: 0.88
+num_questions: 1319
+num_fewshot: 5
+server_args: "--enforce-eager --max-model-len 8192 --data-parallel-size 2 --enable-expert-parallel"
```
Comment (robertgshaw2-redhat) [inline tests/evals/gsm8k/configs/moe-refactor-dp-ep/Qwen3-30B-A3B-BF16-triton.yaml:5]: we should have a flashinfer example here?

Comment (robertgshaw2-redhat) [reply tests/evals/gsm8k/configs/moe-refactor-dp-ep/Qwen3-30B-A3B-BF16-triton.yaml:5]: nvm

**Review by robertgshaw2-redhat — APPROVED** [review]
thanks for adding the CI tests!

File: vllm/model_executor/layers/fused_moe/unquantized_fused_moe_method.py [hunk 11]
```diff
@@ -52,46 +45,17 @@
 logger = init_logger(__name__)
 
 
-# --8<-- [start:unquantized_fused_moe]
```
Comment (bnellnm) [inline vllm/model_executor/layers/fused_moe/unquantized_fused_moe_method.py:55]: I think these weird comments need to preserved for doc purposes.

**Review by mgoin — APPROVED** [review]
